test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java

```diff
@@ -23,216 +23,96 @@
 
 /**
  * @test
- * @bug 8322996
- * @summary Ensure no assert error in C2 with deeply nested synchronize
- *          statements.
- * @run main/othervm -XX:CompileCommand=compileonly,compiler.locks.TestNestedSynchronize::test
+ * @bug 8322996 8324839 8325467
+ * @summary Ensure C2 can compile deeply nested synchronize statements.
+ *          Exercises C2 register masks, in particular. We incrementally
+ *          increase the level of nesting (up to 100) to trigger potential edge
+ *          cases.
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib /
+ * @run main/othervm -XX:CompileCommand=compileonly,Test::test*
  *                   -Xcomp
+ *                   -XX:+UnlockDiagnosticVMOptions
+ *                   -XX:+AbortVMOnCompilationFailure
  *                   compiler.locks.TestNestedSynchronize
  */
 
 package compiler.locks;
 
+import compiler.lib.compile_framework.*;
+import java.util.LinkedList;
+
 public class TestNestedSynchronize {
+    static int min = 1;
+    static int max = 100;
+    static String test_class_name = "Test";
+    static String test_method_name = "test";
 
-    public static void main(String[] args) {
-        test();
+    // The below method generates a program of the form:
+    //
+    // public class Test {
+    //     public static void test1() {
+    //         synchronized (Test.class) {
+    //         }
+    //     }
+    //
+    //     public static void test2() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         }
+    //         }
+    //     }
+    //
+    //     public static void test3() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         }
+    //         }
+    //         }
+    //     }
+    //
+    //     ...
+    //
+    //     public static void test100() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         ...
+    //         }
+    //         }
+    //         }
+    //     }
+    // }
+    //
+    // The above is a massive program. Therefore, we do not directly inline the
+    // program in TestNestedSynchronize and instead compile and run it via the
+    // CompileFramework.
```

Comment:
Nice, I like it. Of course I have no bias here ;)

Reply:
Great :slightly_smiling_face: